---
output: html_document
editor_options: 
  chunk_output_type: console
---
# Code book

```{r, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
## 02 IntroR and RStudio

### File directories
```{r}
getwd()
```

### Running commands and annotating code:

Add some additional code to the following `chunk`:
```{r}
# Intro to R Lesson

3+5
```
The chunk will run inline, but I prefer to set the chunk so it will run in the console, because it gives us extra room, especially for plotting. You can set this using the cog wheel at the top of the script window next to the Knit button.


**Exercise**

Try running only 3 + in the console. Find a way to bring back the command prompt > in the console.

* Ans: 

### The assignment operator (\<-)
The assignment operator assigns values on the right to variables on the left.

**In RStudio the keyboard shortcut for the assignment operator <- is Alt + - (Windows) or Option + - (MacOS).**

```{r}
x <- 3
y <- 5
```

```{r}
number <- x + y
```

**Exercise**

1. Try changing the value of the variable x to 5. What happens to number?
```{r}

```

* Ans:  

2. Now try changing the value of variable y to contain the value 10. What do you need to do, to update the variable number?

* Ans: 

```{r}

```

## 03 IntroR syntax and data structures

### Vectors

```{r}
# Create a numeric vector and store the vector as a variable called 'glengths'
glengths <- c(4.6, 3000, 50000)
glengths
```


```{r}
# Create a character vector and store the vector as a variable called 'species'
species <- c("ecoli", "human", "corn")
species
```

**Exercise**

Try to create a vector of numeric and character values by combining the two vectors that we just created (glengths and species). Assign this combined vector to a new variable called combined. Hint: you will need to use the combine `c()` function to do this. Print the combined vector in the console, what looks different compared to the original vectors?

```{r}

```

* Ans: 

### Factors

```{r}
# Create a character vector and store the vector as a variable called 'expression'
expression <- c("low", "high", "medium", "high", "low", "medium", "high")
expression
```

```{r}

# Turn 'expression' vector into a factor
expression <- factor(expression)
expression
```

**Exercises**

Let’s say that in our experimental analyses, we are working with three different sets of cells: normal, cells knocked out for geneA (a very exciting gene), and cells overexpressing geneA. We have three replicates for each celltype.

1. Create a vector named samplegroup with nine elements: 3 control (“CTL”) values, 3 knock-out (“KO”) values, and 3 over-expressing (“OE”) values.
```{r}

```

2. Turn samplegroup into a factor data structure.

```{r}

```

### Dataframes

A data.frame is the de facto data structure for most tabular data and what we use for statistics and plotting. A data.frame is similar to a matrix in that it’s a collection of vectors of the same length and each vector represents a column. However, in a dataframe each vector can be of a different data type (e.g., characters, integers, factors).

```{r}
# Create a data frame and store it as a variable called 'df'
df <- data.frame(species, glengths)
df
```

**Exercise**

Create a data frame called favorite_books with the following vectors as columns:
```{r}
titles <- c("Catch-22", "Pride and Prejudice", "Nineteen Eighty Four")
pages <- c(453, 432, 328)
```

```{r}

```
### Lists:

```{r}
list1 <- list(species, df, number)
list1
```

**Exercise**

Create a list called list2 containing species, glengths, and number.
```{r}

```

## 04 IntroR: functions and arguments 

### Basic functions

We have already used a few examples of basic functions in the previous lessons i.e `getwd()`, `c()`, and `factor()`.

Here we use `c()`:
```{r}
glengths <- c(glengths, 90) # adding at the end 
glengths
glengths <- c(30, glengths) # adding at the beginning
glengths
```

```{r}
sqrt(81)
```

```{r}
sqrt(glengths)
```

```{r}
round(3.14159)
```

### Seeking help on functions
```{r}
?round
```

```{r}
args(round)
```

```{r}
example("round")
```

```{r}
round(3.14159, digits=2)
```

**Exercise**

1. Let’s use base R function to calculate mean value of the glengths vector. You might need to search online to find what function can perform this task.
```{r}
```

2. Create a new vector test <- c(1, NA, 2, 3, NA, 4). Use the same base R function from exercise 1 (with addition of proper argument), and calculate mean value of the test vector. The output should be 2.5. > NOTE: In R, missing values are represented by the symbol NA (not available). It’s a way to make sure that users know they have missing data, and make a conscious decision on how to deal with it. There are ways to ignore NA during statistical calculation, or to remove NA from the vector. If you want more information related to missing data or NA you can go to the NA help page (please note that there are many advanced concepts on that page that have not been covered in class).
```{r}

```

3. Another commonly used base function is sort(). Use this function to sort the glengths vector in descending order.
```{r}

```

### User-defined functions:

name_of_function <- function(argument1, argument2) {
    |statements or code that does something
    |return(something)
}

This function will take in a numeric value as input, and return the squared value.
```{r}
square_it <- function(x) {
    square <- x * x
    return(square)
}
```

```{r}
square_it(5)
```

**Exercise**

Write a function called multiply_it, which takes two inputs: a numeric value x, and a numeric value y. The function will return the product of these two numeric values, which is x * y. For example, multiply_it(x=4, y=6) will return output 24.
```{r}

```

```{r}

```

## 05 Intro to packages

### Packages and Libraries

You can check what libraries (packages) are loaded in your current R session by typing into the console:
```{r}
sessionInfo() #Print version information about R, the OS and attached or loaded packages

# OR

search() #Gives a list of attached packages
```

### Installing packages

#### Packages from CRAN:

Previously we have introduced you to installing packages. An example is given below for the `ggplot2` package that will be required for some plots we will create later on. Run this code to install `ggplot2`.
 
Set eval = TRUE if you haven't installed ggplot2 yet
```{r, eval=FALSE}
install.packages("ggplot2")
```

Alternatively, packages can also be installed from [Bioconductor](https://www.bioconductor.org/), a  repository of packages which provides tools for the analysis and comprehension of high-throughput **genomic data**. 

Set eval = TRUE if you haven't installed BiocManager yet
```{r, eval=FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(version = "3.17")
```

#### Packages from Bioconductor:

Now you can use BiocManager::install to install packages available on Bioconductor. Bioconductor is a free, open source and open development software project for the analysis and comprehension of genomic data generated by wet lab experiments in molecular biology. 

```{r, eval=FALSE}

# DO NOT RUN THIS!
BiocManager::install("ggplot2")
```

### Loading libraries/packages

```{r}
library(ggplot2)
```

**Exercise**

The ggplot2 package is part of the tidyverse suite of integrated packages which was designed to work together to make common data science operations more user-friendly. We will be using the tidyverse suite in later lessons, and so let’s install it. Use the function from the`BiocManager` package.

Set eval=TRUE if tidyverse is not yet installed. If you've already installed `tidyverse`, how can you run the installation without getting an error?
```{r, eval=FALSE}

```

* Ans: use the argument force = TRUE

## 06 introR: data wrangling

### Vectors: Selecting using indices
```{r}
age <- c(15, 22, 45, 52, 73, 81)
age
age[5]
```

Reverse selection:
```{r}
age[-5]
```

```{r}
age[c(3,5,6)] ## nested

# OR

## create a vector first then select
idx <- c(3,5,6) # create vector of the elements of interest
age[idx]
```

```{r}
age[1:4]
```

**Exercises**

1. Create a vector called alphabets with the following letters, C, D, X, L, F.

```{r}

```

2. Use the associated indices along with [ ] to do the following:

    a) only display C, D and F
  
```{r}

```
    b) display all except X
```{r}

```

   c)  display the letters in the opposite order (F, L, X, D, C)
```{r}

```

### Selecting using logical operators
```{r}
age > 50
```

```{r}
age > 50 | age < 18

age

age[age > 50 | age < 18]  ## nested

# OR

## create a vector first then select
idx <- age > 50 | age < 18
age[idx]
```

### Logical operators with `which()`

```{r}
which(age > 50 | age < 18)

age[which(age > 50 | age < 18)]  ## nested

# OR

## create a vector first then select
idx_num <- which(age > 50 | age < 18)
age[idx_num]
```

### Factors

```{r}
expression <- c("low", "high", "medium", "high", "low", "medium", "high")

expression <- factor(expression)

str(expression)

expression[expression == "high"]    ## This will only return those elements in the factor equal to "high"
```

**Exercise**

Extract only those elements in samplegroup that are not KO (nesting the logical operation is optional).
```{r}


```

#### Releveling factors

```{r}
expression <- factor(expression, levels=c("low", "medium", "high"))     # you can re-factor a factor 

str(expression)
```


**Exercise**

Use the samplegroup factor we created in a previous lesson, and relevel it such that KO is the first level followed by CTL and OE.

```{r}

```

## 07 Reading and data inspection

### Reading data into R:

```{r}
?read.csv
```

```{r}
metadata <- read.csv(file="data/mouse_exp_design.csv")

```

**Exercise 1**

1. Read "project-summary.txt" in to R using read.table() with the approriate arguments and store it as the variable proj_summary. To figure out the appropriate arguments to use with read.table(), keep the following in mind:

* all the columns in the input text file have column name/headers

* you want the first column of the text file to be used as row names       
* hint: look up the input for the row.names = argument in read.table()

```{r}

```

2. Display the contents of proj_summary in your console
```{r}

```

### Inspecting data structures:

Getting info on environmental variables:
```{r}
head(metadata)
str(metadata)
dim(metadata)
nrow(metadata)
ncol(metadata)
class(metadata)
colnames(metadata)
```

**Exercise 2**

Use the class() function on glengths and metadata, how does the output differ between the two?
```{r}

```

Use the summary() function on the proj_summary dataframe, what is the median “Intronic_Rate?

```{r}

```

* Ans: 

How long is the samplegroup factor?

```{r}

```

* Ans:

What are the dimensions of the proj_summary dataframe?

```{r}

```

When you use the rownames() function on metadata, what is the data structure of the output?
```{r}

```

* Ans: 

How many elements in (how long is) the output of colnames(proj_summary)? Don’t count, but use another function to determine this.
```{r}


```

## 08 Data-wrangling: dataframes and lists

### Data wrangling: dataframes

Dataframes (and matrices) have 2 dimensions (rows and columns), so if we want to select some specific data from it we need to specify the “coordinates” we want from it. We use the same square bracket notation but rather than providing a single index, there are two indices required. Within the square bracket, row numbers come first followed by column numbers (and the two are separated by a comma).

Extracting values from data.frames:
```{r}
# Extract value 'Wt'
metadata[1, 1]
```

```{r}
# Extract third row
metadata[3, ] 
```
```{r}
# Extract third column
metadata[ , 3]  
```

```{r}
# Extract third column as a data frame
metadata[ , 3, drop = FALSE] 
```

```{r}
# Dataframe containing first two columns
metadata[ , 1:2] 
```

```{r}
# Data frame containing first, third and sixth rows
metadata[c(1,3,6), ] 
```

```{r}
# Extract the celltype column for the first three samples
metadata[c("sample1", "sample2", "sample3") , "celltype"] 
```

```{r}
# Check column names of metadata data frame
colnames(metadata)

# Check row names of metadata data frame
rownames(metadata)
```

```{r}
# Extract the genotype column
metadata$genotype 
```

```{r}
# Extract the first five values/elements of the genotype column
metadata$genotype[1:5]
```

**Exercises**

Return a data frame with only the genotype and replicate column values for sample2 and sample8.
```{r}

```

Return the fourth and ninth values of the replicate column.
```{r}

```

Extract the replicate column as a data frame.
```{r}

```

#### Selecting using logical operators:

For example, if we want to return only those rows of the data frame with the celltype column having a value of typeA, we would perform two steps:

1. Identify which rows in the celltype column have a value of typeA.

2. Use those TRUE values to extract those rows from the data frame.

To do this we would extract the column of interest as a vector, with the first value corresponding to the first row, the second value corresponding to the second row, so on and so forth. We use that vector in the logical expression. Here we are looking for values to be equal to typeA, 

So our logical expression would be:
```{r}
metadata$celltype == "typeA"
```

```{r}
logical_idx <- metadata$celltype == "typeA"
```

If we want to keep all the rows that have celltype == "typeA":

Any row in the logical_idx that is TRUE will be kept, those that are FALSE will be discarded.
```{r}
metadata[logical_idx, ]
```

```{r}
idx <- which(metadata$celltype == "typeA")
```

```{r}
metadata[idx, ]
```

Let’s try another subsetting. Extract the rows of the metadata data frame for only the replicates 2 and 3. First, let’s create the logical expression for the column of interest (replicate):
```{r}

idx <- which(metadata$replicate > 1)
    
metadata[idx, ]
```

This should return the indices for the rows in the replicate column within metadata that have a value of 2 or 3. Now, we can save those indices to a variable and use that variable to extract those corresponding rows from the metadata table. Alternatively, you could do this in one line:
```{r}
sub_meta <- metadata[which(metadata$replicate > 1), ]
```

**Exercise**

Subset the metadata dataframe to return only the rows of data with a genotype of KO.

```{r}

```

### Selecting components from a list 

Selecting from a list requires slightly different notation, even though in theory a list is a vector (that contains multiple data structures). To select a specific component of a list, you need to use double bracket notation [[]].
```{r}

list1 <- list(species, df, number)
list1[[2]]
```

```{r}
comp2 <- list1[[2]]
class(comp2)
```

```{r}
list1[[1]]
```

```{r}
list1[[1]][1]
```

**Exercises**

Create a list named random with the following components: metadata, age, list1, samplegroup, and number.
```{r}

```

Extract the samplegroup component.
```{r}

```

#### Assigning names to a list

Adding names to components of a list uses the names() function. Let’s check and see if the list1 has names for the components:
```{r}
names(list1)

# Name components of the list
names(list1) <- c("species", "df", "number")
    
names(list1)
```

```{r}
# Extract 'df' component
list1$df
```

**Exercise**

Let’s practice combining ways to extract data from the data structures we have covered so far:

Set names for the random list you created in the last exercise.
```{r}

```

Extract the age component using the $ notation
```{r}

```


## 09 Identifying matching elements

### Logical operators for matching elements

```{r}
rpkm_data <- read.csv("data/counts.rpkm.csv")
```

```{r}
rpkm_data[1:3, 1:3]
ncol(rpkm_data)
nrow(metadata)
```

```{r}
head(rpkm_data)
ncol(rpkm_data)
nrow(metadata)
```

### The `%in%` operator

The `%in%` operator will take each element from vector1 as input, one at a time, and evaluate if the element is present in vector2. The two vectors do not have to be the same size. This operation will return a vector containing logical values to indicate whether or not there is a match. The new vector will be of the same length as vector1. Take a look at the example below:
```{r}

A <- c(1,3,5,7,9,11)   # odd numbers
B <- c(2,4,6,8,10,12)  # even numbers

# test to see if each of the elements of A is in B  
A %in% B
```

Let’s change a couple of numbers inside vector B to match vector A:
```{r}

A <- c(1,3,5,7,9,11)   # odd numbers
B <- c(2,4,6,8,1,5)  # add some odd numbers in 

# test to see if each of the elements of A is in B  
A %in% B
```

```{r}
intersection <- A %in% B
intersection
```

```{r}
A[intersection]
```

```{r}
any(A %in% B)
all(A %in% B)
```

**Exercise** 

1. Using the A and B vectors created above, evaluate each element in B to see if there is a match in A
```{r}


```

2. Subset the B vector to only return those values that are also in A.
```{r}


```

Suppose we had two vectors containing same values. **How can we check if those values are in the same order in each vector?**  In this case, we can use == operator to compare each element of the same position from two vectors. Unlike %in% operator, == operator requires that two vectors are of equal length.
```{r}
A <- c(10,20,30,40,50)
B <- c(50,40,30,20,10)  # same numbers but backwards 

# test to see if each element of A is in B
A %in% B

# test to see if each element of A is in the same position in B
A == B

# use all() to check if they are a perfect match
all(A == B)
```

Let’s try this on our genomic data, and see whether we have metadata information for all samples in our expression data. 
```{r}

x <- rownames(metadata)
y <- colnames(rpkm_data)
```

```{r}
all(x %in% y)
```

```{r}
all(rownames(metadata) %in% colnames(rpkm_data))
```

```{r}
x == y
all(x == y)
```

Looks like all of the samples are there, but they need to be reordered. To reorder our genomic samples, we will learn different ways to reorder data in our next lesson. 

**Exercise**

We have a list of 6 marker genes that we are very interested in. Our goal is to extract count data for these genes using the %in% operator from the rpkm_data data frame, instead of scrolling through rpkm_data and finding them manually.

First, let’s create a vector called important_genes with the Ensembl IDs of the 6 genes we are interested in:

```{r}
important_genes <- c("ENSMUSG00000083700", "ENSMUSG00000080990", "ENSMUSG00000065619", "ENSMUSG00000047945", "ENSMUSG00000081010", "ENSMUSG00000030970")
```

1. Use the %in% operator to determine if all of these genes are present in the row names of the rpkm_data data frame.
```{r}

```

2. Extract the rows from rpkm_data that correspond to these 6 genes using [] and the %in% operator. Double check the row names to ensure that you are extracting the correct rows.
```{r}

```

3. Bonus question: Extract the rows from rpkm_data that correspond to these 6 genes using [], but without using the %in% operator.
```{r}

```

## 10 reordering to match datasets

### Manual reordering using indices

```{r}
teaching_team <- c("Jihe", "Mary", "Meeta", "Radhika", "Will", "Emma")
```

```{r}
# Extracting values from a vector
teaching_team[c(2, 4)] 
```

```{r}

# Extracting all values and reordering them
teaching_team[c(5, 4, 6, 2, 1, 3)]
# Saving the results to a variable
reorder_teach <- teaching_team[c(5, 4, 6, 2, 1, 3)] 

```

**Exercise**

Now that we know how to reorder using indices, let’s try to use it to reorder the contents of one vector to match the contents of another. Let’s create the vectors first and second as detailed below:
```{r}

first <- c("A","B","C","D","E")
second <- c("B","D","E","A","C")  # same letters but different order

```

How would you reorder the second vector to match first?
```{r}

```

### The `match` function


`match()` takes 2 arguments. The first argument is a vector of values in the order you want, while the second argument is the vector of values to be reordered such that it will match the first:

1st vector: values in the order you want
2nd vector: values to be reordered

The `match` function returns the position of the matches (indices) with respect to the second vector, which can be used to re-order it so that it matches the order in the first vector.  Let's use match() on the first and second vectors we created.

```{r}

# Saving indices for how to reorder `second` to match `first`
reorder_idx <- match(first,second) 

```

```{r}

# Reordering the second vector to match the order of the first vector
second[reorder_idx]

```

```{r}

# Reordering and saving the output to a variable
second_reordered <- second[reorder_idx] 

```

Matching with vectors of different lengths:
```{r}

first <- c("A","B","C","D","E")
second <- c("D","B","A")  # remove values

```

```{r}

match(first,second)

```

```{r}

second[match(first, second)]

```

**NOTE: For values that don't match by default return an NA value. You can specify what values you would have it assigned using nomatch argument.**

Example:
```{r}
mtch <- match(first, second, nomatch = 0)
mtch

second[mtch]
```


```{r}

# Check row names of the metadata
rownames(metadata)

# Check the column names of the counts data
colnames(rpkm_data)
```

Reordering genomic data using the `match()` function:
```{r}
genomic_idx <- match(rownames(metadata), colnames(rpkm_data))
genomic_idx
```

```{r}
# Reorder the counts data frame to have the sample names in the same order as the metadata data frame
rpkm_ordered  <- rpkm_data[ , genomic_idx]
```

```{r}
# View the reordered counts
View(rpkm_ordered)

# I prefer to use `head` -- and set the output to a limited number of columns so that it fits in the console

head(rpkm_ordered[,1:5])
```

Now we can check if the rownames of metadata is in the same order as the colnames of rpkm_ordered:
```{r}
all(rownames(metadata) == colnames(rpkm_ordered))
```

**Exercises**

After talking with your collaborator, it becomes clear that sample2 and sample9 were actually from a different mouse background than the other samples and should not be part of our analysis. Create a new variable called subset_rpkm that has these columns removed from the rpkm_ordered data frame.

```{r}

```

Use the `match()` function to subset the metadata data frame so that the row names of the metadata data frame match the column names of the subset_rpkm data frame.
```{r}


```

## 11 Setting up to plot


### Setup a data frame for visualization

```{r}
mean(rpkm_ordered$sample1)
```

### The `map` family of functions

We can use map functions to execute some task/function on every element in a vector, or every column in a dataframe, or every component of a **list**, and so on.

- `map()` creates a list.
- `map_lgl()` creates a logical vector.
- `map_int()` creates an integer vector.
- `map_dbl()` creates a “double” or numeric vector.
- `map_chr()` creates a character vector.

The syntax for the `map()` family of functions is:

- `map`(object, function_to_apply)

We can use the `map_dbl()` to get the mean of each column of `rpkm_ordered` in just one command:
```{r}
# install purrr if you need to by uncommenting the following line:
# BiocManager::install(purrr) 
library(purrr)  # Load the purrr

samplemeans <- map_dbl(rpkm_ordered, mean) 
```


### Adding to a metadata object

```{r}
# Named vectors have a name assigned to each element instead of just referring to them as indices ([1], [2] and so on)
samplemeans

# Check length of the vector before adding it to the data frame
length(samplemeans)
```

```{r}
# Create a numeric vector with ages. Note that there are 12 elements here
age_in_days <- c(40, 32, 38, 35, 41, 32, 34, 26, 28, 28, 30, 32)      
```

Now, we are ready to combine the metadata data frame with the 2 new vectors to create a new data frame with 5 columns:
```{r}
# Add the new vector as the last column to the new_metadata dataframe
new_metadata <- data.frame(metadata, samplemeans, age_in_days) 

# Take a look at the new_metadata object
head(new_metadata)


```

## 12 Data visualization with ggplot2

```{r}
## load the new_metadata data frame into your environment from a .RData object
load("data/new_metadata.RData")
```

```{r}
# this data frame should have 12 rows and 5 columns
head(new_metadata)
```

```{r}
library(ggplot2)
```

```{r}
ggplot(new_metadata) # what happens? we get a plot window
```

```{r, eval=FALSE}
ggplot(new_metadata) +
  geom_point() # note what happens here
```

### Aesthetics

We get an error because each type of geom usually has a required set of aesthetics to 
be set. “Aesthetics” are set with the aes() function and can be set either nested within geom_point() (applies only to that layer) or within ggplot() (applies to the whole plot).

The aes() function has many different arguments, and all of those arguments take columns from the original data frame as input. It can be used to specify many plot elements including the following:

- position (i.e., on the x and y axes)
- color (“outside” color)
- fill (“inside” color)
- shape (of points)
- linetype
- size
  
ggplot(data,aes()) + geoms()
```{r}
ggplot(new_metadata,aes(x = age_in_days, y= samplemeans)) # what happens here? we initialize the plot window with the axes
```

Now add geoms:
```{r}

ggplot(new_metadata,aes(x = age_in_days, y= samplemeans))+geom_point()

#OR 

ggplot(new_metadata) +
     geom_point(aes(x = age_in_days, y= samplemeans))
```

### Geometries

Geoms (e.g.) in layeres:
- geom_point()
- geom_boxplot()
- geom_bar()
- geom_density()
- geom_dotplot()
- geom_label() 
- geom_text()
  
```{r}
ggplot(new_metadata) +
  geom_point(aes(x = age_in_days, y= samplemeans, color = genotype)) 
```

```{r}
ggplot(new_metadata) +
  geom_point(aes(x = age_in_days, y= samplemeans, color = genotype, shape=celltype)) 
```

```{r}
ggplot(new_metadata) +
  geom_point(aes(x = age_in_days, y= samplemeans, color = genotype, shape=celltype), size=2.25) 
```

```{r}
ggplot(new_metadata) +
  geom_point(aes(x = age_in_days, y= samplemeans, color = genotype, shape=celltype), size=3.0) + theme_bw()
```

```{r}
ggplot(new_metadata) +
  geom_point(aes(x = age_in_days, y= samplemeans, color = genotype, shape=celltype), size=2.25) +
  theme_bw() +
  theme(axis.title = element_text(size=rel(1.5)))       
```

**Exercise**

1. The current axis label text defaults to what we gave as input to geom_point (i.e the column headers). We can change this by adding additional layers called xlab() and ylab() for the x- and y-axis, respectively. Add these layers to the current plot such that the x-axis is labeled “Age (days)” and the y-axis is labeled “Mean expression”.

2. Use the ggtitle layer to add a plot title of your choice.

3. Add the following new layer to the code chunk theme(plot.title=element_text(hjust=0.5))

```{r}

```

What does it change?

- Ans: 

How many theme() layers can be added to a ggplot code chunk, in your estimation?

- Ans: 

## 13 Boxplot exercise

Generate a boxplot using the data in the new_metadata dataframe. Create a ggplot2 code chunk with the following instructions:

1. Use the geom_boxplot() layer to plot the differences in sample means between the Wt and KO genotypes.

2. Use the fill aesthetic to look at differences in sample means between the celltypes within each genotype.

3. Add a title to your plot.

4. Add labels, ‘Genotype’ for the x-axis and ‘Mean expression’ for the y-axis.

5. Make the following theme() changes:
    a) Use the theme_bw() function to make the background white.
    b) Change the size of your axes labels to 1.25x larger than the default.
    c) Change the size of your plot title to 1.5x larger than default.
    d) Center the plot title.

1-5.
```{r}

```

6. Change the order of genotype

```{r}

```

7. Change default colors

Add a new layer scale_color_manual(values=c("purple","orange")). Do you observe a change?

- Ans: 

Replace scale_color_manual(values=c("purple","orange")) with scale_fill_manual(values=c("purple","orange")). Do you observe a change?

- Ans: 

In the scatterplot we drew in class, add a new layer scale_color_manual(values=c("purple","orange")), do you observe a difference?

- Ans: 

What do you think is the difference between scale_color_manual() and scale_fill_manual()?

- Ans: 


8. Boxplot using "color" instead of "fill". Use purple and orange for your colors.
```{r}

 
```

9. Back in your boxplot code, change the colors in the scale_fill_manual() layer to be your 2 favorite colors.
```{r}
 
 
```

10. Are there any colors that you tried that did not work?

- Ans: 

11. **OPTIONAL Exercise**

Find the hexadecimal code for your 2 favourite colors (from exercise 3 above) and replace the color names with the hexadecimal codes within the ggplot2 code chunk.
```{r}
# Hint: the gplots package. If gplots is not already installed, uncomment the following line of code

# BiocManager::install("gplots")

library(gplots)

```

```{r}
 
 
```

## 14 Exporting data and plots

### Writing data to file

```{r}
# Save a data frame to file
write.csv(sub_meta, file="data/subset_meta.csv")
```

```{r}
# Save a vector to file
write(glengths, file="data/genome_lengths.txt")
```

Note, the ncolumns argument that it defaults to five columns unless specified, so to get a single column:

```{r}
# Save a vector to file as a single column
write(glengths, file="data/genome_lengths.txt", ncolumns = 1)
```

### Exporting figures to file:

```{r}
## Make a plot which will be written to the open device, in this case the temp file created by pdf() or png()
##  
p <- ggplot(new_metadata) +
  geom_point(aes(x = age_in_days, y= samplemeans, color = genotype, shape=celltype), size=rel(3.0)) 
```

```{r}

## Open device for writing
pdf(file = "figures/scatterplot.pdf")
print(p)

# OR

pdf(file = "figures/scatterplot.pdf")
ggplot(new_metadata) +
  geom_point(aes(x = age_in_days, y= samplemeans, color = genotype,
            shape=celltype), size=rel(3.0)) 

```

```{r}
## Closing the device is essential to save the temporary file created by pdf()/png()
dev.off()
```

Alternative method with ggplot2:
```{r}

ggplot(new_metadata) +
  geom_point(aes(x = age_in_days, y= samplemeans, color = genotype, shape=celltype), size=rel(3.0)) 

ggsave("figures/scatterplot.pdf")

```

## 15 Finding help

### Saving variables

In order to find help on line you must provide a reproducible example. **Google** and R sites such as **Stackoverflow** are good sources. To provide an example you may want to share an object with someone else, you can save any or all R data structures that you have in your environment to a file:
```{r,eval=FALSE}

save.image() # saves all environmental variables in your current R session to a file called ".RData" in your working directory -- this is a command that you should use often when in an R session to save your work.

#OR give it a memorable name

save.image("intro-to-R.RData")

#OR for a single object

save.image(metadata,"metadata.RData")

# you can also use 

saveRDS(metadata,file = "metadata.rds")

```

### Loading data
```{r, eval=FALSE}

load(file = ".RData") # to load .RData files

readRDS(file = "metadata.rds") # to load .rds files

```

### `sessionInfo()`

Lists details about your current R session: R version and platform; locale and timezone; all packages and versions that are loaded as we saw before.
```{r, eval=FALSE}
# you'll also often be required to provide your session info
sessionInfo()
```

## 16 Data wrangling with Tidyverse

Set message to FALSE in a chunck if you want to avoid the messages when loading libraries
```{r, message=FALSE}
# If tidyverse is not already installed, uncomment the following line of code
# BiocManager::install("tidyverse")

library(tidyverse)
```

**Tidyverse basics**

### **Pipes: The pipe (`%>%`) allows the output of a previous command to be used as input to another command instead of using nested functions.**

**NOTE: Shortcut to write the pipe is shift + command + m on Macos; shift + ctrl + m on Windows**

```{r}
## A single command
sqrt(83)

## Base R method of running more than one command
round(sqrt(83), digits = 2)

## Running more than one command with piping
sqrt(83) %>% round(digits = 2)
```

**Exercise**

1. Create a vector of random numbers using the code below:

```{r}

```

2. Use the pipe (%>%) to perform two steps in a single line: 

    a) Take the mean of random_numbers using the mean() function.
```{r}

```
    
    b)  Round the output to three digits using the round() function.
```{r}

```

### Tibbles

**Experimental data**

The dataset: gprofiler_results_Mov10oe.tsv

The gene list represents the functional analysis results of differences between control mice and mice over-expressing a gene involved in RNA splicing. We will focus on gene ontology (GO) terms, which describe the roles of genes and gene products organized into three controlled vocabularies/ontologies (domains):

- biological processes (BP)

- cellular components (CC)

- molecular functions (MF) 

that are over-represented in our list of genes. 

### Analysis goal and workflow

**Goal:** *Visually compare the most significant biological processes (BP) based on the number of associated differentially expressed genes (gene ratios) and significance values and create a plot.*

**1. Read in the functional analysis results**
```{r}

# Read in the functional analysis results
functional_GO_results <- read.delim(file = "data/gprofiler_results_Mov10oe.tsv")

# Take a look at the results
functional_GO_results[1:3,1:12]

```

```{r}

names(functional_GO_results)
```

**2. Extract only the GO biological processes (BP) of interest**
```{r}

# Return only GO biological processes
bp_oe <- functional_GO_results %>%
  filter(domain == "BP")
  
bp_oe[1:3,1:12]
```

**Exercise**

We would like to perform an additional round of filtering to only keep the most specific GO terms.

1. For bp_oe, use the filter() function to only keep those rows where the relative.depth is greater than 4.
2. Save output to overwrite our bp_oe variable.
```{r}

```

### Select

3. Select only the colnames needed for visualization
```{r}

# Selecting columns to keep for visualization
names(bp_oe)
bp_oe <- bp_oe %>%
  select(term.id, term.name, p.value, query.size, term.size, overlap.size, intersection)

head(bp_oe[,1:6])

dim(bp_oe)
```

The `select()` function also allows for negative selection. So we could have alternately removed columns with negative selection. Note that we need to put the column names inside of the combine (`c()`) function with a - preceding it for this functionality.

```{r, eval=FALSE}

# DO NOT RUN
# Selecting columns to remove
bp_oe <- bp_oe %>%
    select(-c(query.number, significant, recall, precision, subgraph.number, relative.depth, domain))
```

### Arrange

Let’s sort the rows by adjusted p-value with the arrange() function.

```{r}

# Order by adjusted p-value ascending
bp_oe <- bp_oe %>%
  arrange(p.value)
```

**NOTE: If you wanted to arrange in descending order, then you could have run the following instead:**
```{r, eval=FALSE}

# DO NOT RUN
# Order by adjusted p-value descending
bp_oe <- bp_oe %>%
  arrange(desc(p.value))
```

### Rename

Let’s rename the term.id and term.name columns.

```{r}

# Provide better names for columns
bp_oe <- bp_oe %>% 
  dplyr::rename(GO_id = term.id, 
                GO_term = term.name)

```

**Exercise**

Rename the intersection column to genes to reflect the fact that these are the DE genes associated with the GO process.
```{r}

```

### Mutate

Create additional metrics for plotting (e.g. gene ratios)
```{r}
# Create gene ratio column based on other columns in dataset
bp_oe <- bp_oe %>%
  mutate(gene_ratio = overlap.size / query.size)
```

**Exercise**

Create a column in bp_oe called term_percent to determine the percent of DE genes associated with the GO term relative to the total number of genes associated with the GO term (overlap.size / term.size)

```{r}

```

**Exercise**

Use ggplot to plot the first 30 significant GO terms vs gene_ratio. Color the points by p.value as shown in the figure in 16_tidyverse.md.
```{r}
sub = bp_oe[1:30,]

ggplot(sub, aes(x = p.value, y = GO_term, col = p.value)) + geom_point()

ggplot(sub,aes(x = gene_ratio,y = GO_term, color = p.value)) + geom_point() + theme_bw() 
```

Change the p.value gradient colors to red-orange-yellow
```{r}
# to help linearize the p.values for the colors; not essential for answer
sub$`-log10(p.value)` = -log10(sub$p.value) 

g <- ggplot(sub,aes(x = gene_ratio,y = GO_term, color = -log10(p.value))) + geom_point() + theme_bw()

mid <- mean(sub$`-log10(p.value)`)

g + scale_color_gradient2(midpoint = mid, low = "yellow", mid = "orange",high = "red" )

```

